### ISSUE TITLE
Infinite loop in read_cb() SecureTransport callback for SSL support on OS X

### ISSUE THREAD

**Issue by swisspol** [libgit2/libgit2#3195]
```c
static OSStatus read_cb(SSLConnectionRef conn, void *data, size_t *len)
{
	git_stream *io = (git_stream *) conn;
	ssize_t ret;
	size_t left, requested;

	requested = left = *len;
	do {
		ret = git_stream_read(io, data + (requested - left), left);
		if (ret < 0) {  <----------- BUG
			*len = 0;
			return -1;
		}

		left -= ret;
	} while (left);

	*len = requested;

	if (ret == 0)
		return errSSLClosedGraceful;

	return noErr;
}
```

If `git_stream_read()` returns 0, the loop is never broken.

PS: This has been observed at least once in the field: http://forums.gitup.co/t/cpu-usage-went-up-to-100/172.

**Comment by swisspol** [comment]
The reason I'm not submitting a PR for this although it seems an obvious is I'm not sure what the behavior should be if `git_stream_read()` returns > 0 on the first call and then 0 on the following one. Should the function return `errSSLClosedGraceful` or `noErr`?

**Comment by swisspol** [comment]
@carlosmn I guess this one is for you since you added SecureTransport support to libgit2.

**Comment by swisspol** [comment]
I put a fix in #3208.
